File: src/terminal.c
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <termios.h>
#include "eko.h"
#include "terminal.h"

void clear_screen(void) {
    write(STDOUT_FILENO, "\x1b[2J", 4);
    write(STDOUT_FILENO, "\x1b[H", 3);
}

void die(const char *s) {
    clear_screen();
    perror(s);
    exit(1);
}

void disable_raw_mode(void) {
    if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &editor.default_termios) == -1)
        die("tcsetattr");
}

void enable_raw_mode(void) {
    if (tcgetattr(STDIN_FILENO, &editor.default_termios) == -1) die("tcgetattr");
    atexit(disable_raw_mode);

    struct termios raw = editor.default_termios;
    raw.c_iflag &= ~(BRKINT | INPCK | ISTRIP | ICRNL | IXON);
    raw.c_oflag &= ~(OPOST);
    raw.c_cflag &= ~(CS8);
    raw.c_lflag &= ~(ECHO | ICANON | ISIG | IEXTEN);
    raw.c_cc[VMIN] = 0;
    raw.c_cc[VTIME] = 1;

    if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw) == -1) die("tcsetattr");
}

int get_cursor_position(int *rows, int *cols) {
    char buf[32];
    unsigned int i = 0;
    if (write(STDOUT_FILENO, "\x1b[6n", 4) != 4) return -1;
    while (i < sizeof(buf) - 1) {
        if (read(STDIN_FILENO, &buf[i], 1) != 1) break;
        if (buf[i] == 'R') break;
        i++;
    }
    buf[i] = '\0';
    if (buf[0] != '\x1b' || buf[1] != '[') return -1;
    if (sscanf(&buf[2], "%d;%d", rows, cols) != 2) return -1;
    return 0;
}

int get_window_size(int *rows, int *cols) {
    struct winsize ws;
    if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &ws) == -1 || ws.ws_col == 0) {
        if (write(STDOUT_FILENO, "\x1b[999C\x1b[999B", 12) != 12) return -1;
        return get_cursor_position(rows, cols);
    }
    *cols = ws.ws_col;
    *rows = ws.ws_row;
    return 0;
}

File: src/terminal.zig
const std = @import("std");
const posix = std.posix;

pub const Size = struct { rows: u16, cols: u16 };

var default_termios: posix.termios = undefined;

pub fn write(msg: []const u8) !void {
    const stdout = std.fs.File.stdout();
    try stdout.writeAll(msg);
}

pub fn clearScreen() !void {
    write("\x1b[2J\x1b[H") catch {};
}

pub fn die(msg: []const u8) noreturn {
    clearScreen() catch {};
    write(msg) catch {};
    std.process.exit(1);
}

pub fn enableRawMode() !void {
    default_termios = try posix.tcgetattr(posix.STDIN_FILENO);

    var raw = default_termios;

    raw.iflag.BRKINT = false;
    raw.iflag.INPCK = false;
    raw.iflag.ISTRIP = false;
    raw.iflag.ICRNL = false;
    raw.iflag.IXON = false;

    raw.oflag.OPOST = false;

    raw.cflag.CSIZE = .CS8;

    raw.lflag.ECHO = false;
    raw.lflag.ICANON = false;
    raw.lflag.ISIG = false;
    raw.lflag.IEXTEN = false;

    raw.cc[@intFromEnum(posix.V.MIN)] = 0;
    raw.cc[@intFromEnum(posix.V.TIME)] = 1;

    try posix.tcsetattr(posix.STDIN_FILENO, .FLUSH, raw);
}

pub fn disableRawMode() void {
    posix.tcsetattr(posix.STDIN_FILENO, .FLUSH, default_termios) catch {};
}

pub fn getCursorPosition() !Size {
    var buf: [32]u8 = undefined;
    var i: u32 = 0;

    try write("\x1b[6n");

    while (i < buf.len) {
        const n = try std.posix.read(std.posix.STDIN_FILENO, buf[i .. i + 1]);
        if (n == 0) return error.EndOfStream;
        if (buf[i] == 'R') break;
        i += 1;
    }

    if (buf[0] != '\x1b' or buf[1] != '[') return error.InvalidPosition;
    const payload = buf[2..i];

    var it = std.mem.splitScalar(u8, payload, ';');
    const rows = try std.fmt.parseInt(u16, it.next() orelse return error.InvalidPosition, 10);
    const cols = try std.fmt.parseInt(u16, it.next() orelse return error.InvalidPosition, 10);

    return Size{ .rows = rows, .cols = cols };
}

pub fn getWindowSize() !Size {
    var winsize: posix.winsize = undefined;

    const err = posix.system.ioctl(posix.STDOUT_FILENO, posix.T.IOCGWINSZ, @intFromPtr(&winsize));
    if (posix.errno(err) != .SUCCESS) {
        return error.PosixFailure;
    }

    return Size{ .rows = winsize.row, .cols = winsize.col };
}
